design/controlling-agent-updates.md

```diff
@@ -0,0 +1,66 @@
+# Controlling agent updates
+
+The agent is considered an extension of the service.
+As such, we have an auto-update mechanism to ensure that new features can rely on the agent support they require.
+Conceptually, it's no different from binaries on the app tier or JavaScript served to the web frontend.
+
+For instance, tasks express a minimum agent version required.
+If a selected agent doesn't meet that requirement, the agent is first asked to update itself.
+Other ways an agent can be updated include using a feature which demands an evergreen agent (YAML pipelines, for instance) or clicking the "upgrade all agents" button in a pool.
+
+For some highly-regulated customers, this kind of evergreen infrastructure is at odds with their compliance needs.
+Such customers today are doing things like ACLing the agent's installation directory so it can't update itself or recompiling the agent without update support.
+This leaves us in a situation where no one is happy:
+we can't service the product correctly, and the customer has to maintain custom configuration / infrastructure to meet their compliance needs.
+
+## Solution
+
+_See below for other solutions considered._
+
+We will add a mode to the project-level pool which puts it into "no update" mode.
+In this mode, jobs will use the minimum agent version as a demand.
+If no agent can satisfy the demand, the job will fail instead of forcing an agent update.
+
+The error message must clearly indicate what feature or task demanded agent update.
+It must also indicate that the customer opted into this behavior.
+This way, the customer can decide whether to update their agents or edit the pipeline to remove the offending feature.
+For example:
+> This pool doesn't allow agent updates.
+> An administrator can alter this setting on the project pool page.
+> Task `MyTask` version 1.2.3 requires agent version 2.175.1, which could not be satisfied.
+
+Manual agent updates (i.e. clicking the button in the web) should still send update requests as normal.
```

Comment:
This update should be scoped to the pool. Customers may need to control which version it gets updated to. E.g. customer takes 3 weeks to validate or certify a particular agent, and in the mean time we release a new version. Customer will want to upgrade to the certified version. That starts to get pretty complicated!

Reply:
Yep, those updates are already pool-scoped.

Reply:
@vtbassmatt I think you missed my other comment. Do we have a requirement that the customer can control which version of the agent they update to? See scenario above, where the customer will have validated or certified a specific version of the agent, and we may have shipped new agents in the meantime. Or is the customer responsible getting the new agent bits on the agent machine?

Reply:
Given that the primary driver of this feature is people who want to explicitly control agent updates, even with the human overhead that implies, I think the simpler thing is to disallow manual updates altogether when this feature is enabled. That avoids the complexity of evolving the UI to let you dictate which version you want to push down (to @edglas1's point). If this manual update option still exists for a "locked down" pool, we'll also need some way of auditing who has used it, when someone invariably updates a pool outside of their org's normal policy.

Reply:
That makes sense to me. Do we provide any facility for updating the agent? I wonder how Skype does it. They must have 100s of agents.